### PR TITLE
Support async_count for scopes

### PR DESCRIPTION
**Pull request by jdlubrano**
# Background

In ActiveAdmin applications, a surprising performance bottleneck can be the `COUNT` queries that run to provide counts for each scope made available on an resource's index page. In many cases, performance issues can be addressed by adding missing indices and/or otherwise speeding up the underlying query.

There is another performance boost that could be low-hanging fruit, however. The `COUNT` queries for scopes currently run sequentially.  In Rails 7.1, the framework provides the `async_count` method that allows queries to be run in parallel. For index pages with many scopes (and many COUNTs) running queries in parallel could offer a significant time savings.

This PR will use `async_count` to query the collection size of scopes in versions of Rails where `async_count` is available. Rails apps not running on Rails 7.1 will continue to rely on sequential queries.

# Approach

This PR introduces an `ActiveAdmin::AsyncCount` class that quacks enough like an ActiveRecord Relation to play nicely with the `collection_size` helper.  The `ActiveAdmin::Views::Scopes` class eagerly initializes `AsyncCount`s for each scope that is configured like
```ruby
scope(:my_scope, show_count: :async)
```
which kicks off async queries.  As a result, rendering the scopes component should only be as slow as its slowest `COUNT`.

## Closed Questions

* Should the async query behavior be more opt-in? ✅ (it was made more opt-in in the latest commits)
  * ~Should individual scopes have an `async_count: true` option?~ (I choose `show_count: :async` to leverage the existing `show_count` attribute)
  * ~Should the index options support a `scope_count: :async` option?~ (It seems easy enough to configure many scopes to use `async_count` using the `with_options` Rails/ActiveSupport helper.  Example:
```ruby
with_options(show_count: :async) do
  scope(:one)
  scope(:two)
end
```
  * ~Should consumers be able to configure their own `ScopeCounter` class globally and/or per index?~ Not necessary

# Credits

@tuffylock @aturar

### CODE REVIEW

File: lib/active_admin/async_count.rb
Comment (jdlubrano) [inline lib/active_admin/async_count.rb:12]: I'm not entirely sure what a "good" solution is to this.  Rails 7.1.3 has a bug with its query cache such that promises get wrapped by an "extra" promise.  It doesn't seem like ActiveAdmin's responsibility to work around that, but I'm not sure what alternatives there might be.  In any case, this is isolated to this class and is tested, so that seemed like a good starting point.

**Comment by jdlubrano** [comment]
Hi @javierjulio.  I was wondering if there was any interest in this PR from the ActiveAdmin maintainers.  If not, I have no issue closing it.  Sorry about the direct ping, I was looking for a maintainer to nudge.

**Comment by javierjulio** [comment]
@jdlubrano thank you. Just have to be patient for us to find time.

I agree this can be a helpful feature. I haven't tried the new async querying capability in Rails yet so I'll have to rely on others for input. Looking at the scopes DSL in our docs, it seems best that this would be an option per scope and that it should be an opt-in feature.

I'm not sure about your last question though. What need do you see for having the ability to override the class?

**Comment by jdlubrano** [comment]
Thanks so much for the feedback!

I will work on tweaking this to rely on a per-scope, opt-in action. Given the per-scope option, I don't see a need to leave the class open to overrides.

**Comment by jdlubrano** [comment]
I have updated scopes to accept an option like
```ruby
scope(:my_scope, show_count: :async)
```
as a way to opt-in to async queries.  If someone uses `show_count: :async` with a version of Rails/ActiveRecord that does not implement `async_count`, a `NoMethodError` will be raised.  Since this behavior is opt-in, that seems okay to me.

The PR description has also been updated accordingly.

File: lib/active_admin/views/components/scopes.rb
Comment (mgrunberg) [inline lib/active_admin/views/components/scopes.rb:25]: I would keep a single iteration. Performance is not the main argument because I belive scopes is a small collection. But by introducing a select before the each iteration it's not so evident that the main change the introduction of the new helper `display_scope?`

```suggestion
            group_scopes.each do |scope|
              build_scope(scope, options) if display_scope?(scope)
            end
```

File: spec/unit/views/components/scopes_spec.rb
Comment (mgrunberg) [inline spec/unit/views/components/scopes_spec.rb:72]: Would be easy to display a nicer error message in this scenario? 
Something like "scope `show_count: :async` is not available in Rails X.Y.Z. This Rails version doesn't support async_count". 
I know this feature is `opt-in` but I'm wondering if we can explain in a better way the cause of the problem. I want to avoid future issues asking why they get NoMethodError.
Of course, if it's easy.

File: lib/active_admin/async_count.rb
Comment (mgrunberg) [reply lib/active_admin/async_count.rb:12]: Implementing the workaround here makes sense. We don't know the Rails version the host app is running.

**Review by mgrunberg — COMMENTED** [review]
@jdlubrano thanks for your work. I like the PR feature.
I wrote a few comments. I want to know your thoughts about them.

I hope I have more time tomorrow to keep reviewing the PR.

File: spec/unit/views/components/scopes_spec.rb
Comment (jdlubrano) [reply spec/unit/views/components/scopes_spec.rb]: Done in [f4205c3](https://github.com/activeadmin/activeadmin/pull/8394/commits/f4205c30a48a531a627b639bb34d12f8c4870fb1)

File: lib/active_admin/views/components/scopes.rb
Comment (jdlubrano) [reply lib/active_admin/views/components/scopes.rb]: Done in [7dd372e](https://github.com/activeadmin/activeadmin/pull/8394/commits/7dd372e8beffdf498d181375459dc7689a67e093)

**Review by javierjulio — COMMENTED** [review]
@mgrunberg I think this is ready and looks good to me. Would you be able to test it out?

**Review by mgrunberg — COMMENTED** [review]
@javierjulio looks good to me.

I tested it on my machine and it works. I have a simple suggestion before merging.

File: lib/active_admin/views/components/scopes.rb
Comment (jdlubrano) [reply lib/active_admin/views/components/scopes.rb]: Done in 8b044d9da1390c17979c40070f1fd4ef440eb2e7

File: spec/unit/async_count_spec.rb
Comment (jdlubrano) [reply spec/unit/async_count_spec.rb]: Done in be9e118a57f1d1b9886b7e3ac9481e8d4bb7d230

**Review by mgrunberg — APPROVED** [review]

**Review by javierjulio — APPROVED** [review]
Thanks!

**Pull request merged**
